core/src/cluster_nodes.rs

```diff
@@ -211,6 +229,67 @@ fn get_nodes(cluster_info: &ClusterInfo, stakes: &HashMap<Pubkey, u64>) -> Vec<N
     .collect()
 }
 
+impl<T> ClusterNodesCache<T> {
+    pub(crate) fn new(
+        // Capacity of underlying LRU-cache in terms of number of epochs.
+        cap: usize,
+        // A time-to-live eviction policy is enforced to refresh entries in
+        // case gossip contact-infos are updated.
+        ttl: Duration,
+    ) -> Self {
+        Self {
+            cache: Mutex::new(LruCache::new(cap)),
+            ttl,
+        }
+    }
+}
+
+impl<T: 'static> ClusterNodesCache<T> {
+    fn get_cache_entry(&self, epoch: Epoch) -> Arc<Mutex<CacheEntry<T>>> {
+        let mut cache = self.cache.lock().unwrap();
+        match cache.get(&epoch) {
+            Some(entry) => Arc::clone(entry),
+            None => {
+                let entry = Arc::default();
+                cache.put(epoch, Arc::clone(&entry));
+                entry
+            }
+        }
+    }
+
+    pub(crate) fn get(
+        &self,
+        shred_slot: Slot,
+        root_bank: &Bank,
+        cluster_info: &ClusterInfo,
+    ) -> Arc<ClusterNodes<T>> {
+        let epoch = root_bank.get_leader_schedule_epoch(shred_slot);
+        let entry = self.get_cache_entry(epoch);
+        // Hold the lock on the entry here so that, if needed, only
+        // one thread recomputes cluster-nodes for this epoch.
+        let mut entry = entry.lock().unwrap();
+        if let Some((asof, nodes)) = entry.deref() {
+            if asof.elapsed() < self.ttl {
+                return Arc::clone(nodes);
+            }
+        }
+        let epoch_staked_nodes = root_bank.epoch_staked_nodes(epoch);
+        if epoch_staked_nodes.is_none() {
+            inc_new_counter_info!("cluster_nodes-unknown_epoch_staked_nodes", 1);
+            if epoch != root_bank.get_leader_schedule_epoch(root_bank.slot()) {
+                return self.get(root_bank.slot(), root_bank, cluster_info);
+            }
+            inc_new_counter_info!("cluster_nodes-unknown_epoch_staked_nodes_root", 1);
+        }
```

Comment:
If this function were to return an option, I think it's simpler to return `None` here and drop retransmit of a packet if the epoch stakes are so far ahead that we don't have an entry in the root bank for the epoch yet. 

Also, we already filter out these types of shreds that are too far ahead of the leader schedule in retransmit since we cannot determine the leader for signature verification, and it's good avoid retransmitting these types of shreds in case it's some sort of DOS vector anyways

Reply:
> Can you please point me to where in the code-base above "2 epoch ahead" is implemented?
I find below in bank::new_from_parent:
https://github.com/solana-labs/solana/blob/cde146155/runtime/src/bank.rs#L1293-L1294
but this seems to be only one epoch ahead.

Ha you don't need me, you hit it right on the nose! Yeah it's 1 epoch, I'm misremembering information from the 🦖 times.

> Should we have fallback here that if the root-bank did not have the stakes for the epoch, then use the working-bank

I think we should fail here by returning `None`. If we are an epoch behind then we can't make an accurate decision so I think it's better to let the caller handle this

Reply:
> If we are an epoch behind then we can't make an accurate decision so I think it's better to let the caller handle this

But it is not actually that we are behind. In the example I gave above the node was actually catching up with the cluster. But at the epoch boundary, it does not have epoch-stakes for the needed epoch.

The current master code is using:
```rust
bank.get_leader_schedule_epoch(shred_slot)
```
which is **one epoch after**
```rust
epoch_schedule.get_epoch(shred_slot)
```
And at the epoch boundary shred is already one epoch after the root-slot. So in aggregate it will be 2 epochs ahead.

Reply:
> And at the epoch boundary shred is already one epoch after the root-slot. So in aggregate it will be 2 epochs ahead.

Ah I see, that's gross, thanks for the explanation. 

My mental model was flawed because I didn't realize we were using the `leader_schedule_epoch` for the given slot to determine the `epoch_slots`: https://github.com/solana-labs/solana/blob/v1.7/core/src/broadcast_stage/standard_broadcast_run.rs#L225. This makes me wonder why we're not just using the epoch of the slot itself, i.e. `bank.epoch_schedule.get_epoch(slot)`

Reply:
> This makes me wonder why we're not just using the epoch of the slot itself, i.e. bank.epoch_schedule.get_epoch(slot); the epoch of the current slot should already be fixed in broadcast, and this should reduce the gap between the root bank and the slot to 1 epoch even on epoch boundaries.

Agree. I do not know if there is some specific reasons (or some corner cases) that requires to use `get_leader_schedule_epoch` instead. :thinking: 

> Rather than gambling with another layer of potential inconsistency, should we featurize a change to change the cluster to use bank.epoch_schedule.get_epoch(slot) instead of the bank.get_leader_schedule_epoch(slot) first, and then push this change?

I think if we just fallback on the working-bank here it will be backward compatible because that is already what the current master code is doing (i.e. it uses working-bank). I added this logic and I feel like we should be safe to merge this now (again due to backward compatibility). I did some further testing as well on test gce clusters.

Also, as you mentioned, earlier in the pipeline the shreds are sig-verified by the known leader of the slot. So it has already passed some sanity checks.

Once this is merged in, we can investigate if it makes sense to switch to `get_epoch` instead and remove working-bank entirely.

Reply:
> I think if we just fallback on the working-bank here it will be backward compatible because that is already what the current master code is doing

Yeah the only difference I was thinking of was when the working bank crossed the epoch boundary into the next epoch, then we got a shred `S` from the old epoch.

Then in the old code we would be using the stakes from the next epoch
In this new code we would be using the stakes based on the old epoch for shred `S`

Perhaps this is a small enough window where things are relatively fine?

Reply:
That is actually an inconsistency that this commit is intentionally trying to remove.
Please see "1." in the pull request description.

Reply:
Right, but was more worried about the upgrade between people with and without this code change.

Reply:
We can monitor on testnet first and see how it goes. I am expecting it to be fine as the fanout is pretty large.